include/albatross/src/utils/linalg_utils.hpp

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2019 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef ALBATROSS_UTILS_LINALG_UTILS_HPP_
+#define ALBATROSS_UTILS_LINALG_UTILS_HPP_
+
+namespace albatross {
+
+namespace details {
+
+template <typename FeatureType, typename Comparator>
+inline void _print_eigen_directions(const Eigen::MatrixXd &matrix,
+                                    const std::vector<FeatureType> &features,
+                                    std::size_t count, Comparator comp,
+                                    std::ostream *stream) {
+  Eigen::SelfAdjointEigenSolver<Eigen::MatrixXd> ev(matrix);
+  const Eigen::VectorXd values = ev.eigenvalues().real();
+  const Eigen::MatrixXd vectors = ev.eigenvectors();
+
+  using ValueAndVector = std::tuple<double, Eigen::VectorXd>;
+  std::vector<ValueAndVector> values_and_vectors;
+
+  for (Eigen::Index i = 0; i < values.size(); ++i) {
+    values_and_vectors.emplace_back(values[i], vectors.col(i));
+  }
+
+  std::sort(values_and_vectors.begin(), values_and_vectors.end(),
+            [&](const ValueAndVector &a, const ValueAndVector &b) -> bool {
+              return comp(std::get<0>(a), std::get<0>(b));
+            });
+
+  (*stream) << "MIN: " << values.minCoeff() << std::endl;
+  (*stream) << "MAX: " << values.maxCoeff() << std::endl;
+
+  for (std::size_t i = 0; i < count; ++i) {
+    const double value = std::get<0>(values_and_vectors[i]);
+    const auto vector = std::get<1>(values_and_vectors[i]);
+    (*stream) << "eigen value: " << value << std::endl;
+    for (Eigen::Index j = 0; j < vector.size(); ++j) {
+      double coef = vector[j];
+      if (fabs(coef) > 1e-3) {
+        (*stream) << "    " << std::setw(12) << coef << "   " << features[j]
+                  << std::endl;
+      }
```

Comment:
Can you explain the `1e-3` cutoff?
Do you have an example of what this looks like when printed? It looks like it would print  each feature a number of times equal to the height of the matrix.

If an element of an eigenvector is small, print the feature that corresponds to that dimension. Can you explain how this could be used to see unobservable and over constrained features alike?

Reply:
Yeah, 1e-3 was a pretty arbitrary pick, but mostly just a way of removing smaller influences.  I'll change it to an argument.

Reply:
There's a test case which when printed looks like this:
```
MIN: 0.0344498
MAX: 8.74038
eigen value: 0.0344498
        0.174948   0
       -0.690045   1
       0.0589917   2
       -0.314007   3
      -0.0752929   4
      -0.0982879   5
         0.24261   6
         0.45257   7
       -0.254815   8
        -0.21728   9
eigen value: 0.292064
        0.332327   0
       0.0426091   1
       -0.267221   2
...
```
Not a super informative example, but shows the printing format, in this case the features are `int`.  I'll include another example in the description here for future reference.

Reply:
That was an excellent description. Thank you.